### PR TITLE
MySQL: parser v2

### PR DESCRIPTION
**Pull request by QianKaiLin**
Make sure these boxes are checked accordingly before submitting your Pull Request -- thank you.

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [x] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [x] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)

Link to ticket: https://github.com/OISF/suricata-verify/pull/2067

Describe changes:

- fix mysql parser bug.

- fix mysql logger bug.

- add mysql detection keywords `mysql.command` and `mysql.rows`

- add SV tests.

SV_REPO=
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2067

### CODE REVIEW

**Review by catenacyber — COMMENTED** [review]
Thanks for the work so far, still more to do

CI : 🔴 could you fix it ?
Commits segmentation : I think we will squash all in the end, especially the commits like ` mysql: fix`, so we do not have a buggy point in our git history if we can avoid it
Commit messages : ok, we can see after squashing
Git ID set : ok
CLA : ❓ I do not have access
Doc update : 🟠 I think there is more to do, like a note in upgrade guide cc @jufajardini 
Redmine ticket : ok
Rustfmt : ok
Tests : thanks for the SV test
Dependencies added: none
Code : will look into it

**Review by catenacyber — CHANGES_REQUESTED** [review]
CI is to be fixed

File: etc/schema.json [hunk 1]
```diff
@@ -2383,6 +2383,31 @@
             },
             "additionalProperties": false
         },
+        "mysql": {
+            "type": "object",
+            "optional": true,
+            "properties": {
+                "version": {
+                    "type": "string"
```
Comment (catenacyber) [inline etc/schema.json:2391]: Could you add `description` for these fields ?

File: suricata.yaml.in [hunk 2]
```diff
@@ -916,6 +917,12 @@ app-layer:
       stream-depth: 0
       # Maximum number of live PostgreSQL transactions per flow
       # max-tx: 1024
+    mysql:
+      enabled: no
+      # Stream reassembly size for MySQL. By default, track it completely.
+      stream-depth: 0
```
Comment (catenacyber) [inline suricata.yaml.in:923]: Why that default ?

File: src/detect-mysql-rows.c [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        return NULL;
+    }
+    InspectionBufferSetupMulti(buffer, transforms, data, data_len);
+    buffer->flags = DETECT_CI_FLAGS_SINGLE;
+
+    SCReturnPtr(buffer, "InspectionBuffer");
+    return buffer;
+}
+
+void DetectMysqlRowsRegister(void)
```
Comment (catenacyber) [inline src/detect-mysql-rows.c:72]: You should be able to do these keywords in rust only ?
You can look at sip for a latest translation from C to rust for example ;-)

File: rust/src/mysql/logger.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    tx: *mut std::os::raw::c_void, js: &mut JsonBuilder,
+) -> bool {
+    let tx_mysql = cast_pointer!(tx, MysqlTransaction);
+    SCLogDebug!(
+        "----------- MySQL rs_mysql_logger call. Tx is {:?}",
+        tx_mysql
+    );
+    let result = log_mysql(tx_mysql, js);
+    if let Err(ref err) = result {
+        SCLogError!("----------- MySQL rs_mysql_logger failed. err is {:?}", err);
```
Comment (catenacyber) [inline rust/src/mysql/logger.rs:66]: I am not sure we would want such logs in the end

File: rust/src/mysql/logger.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+// written by linqiankai <linqiankai@geweian.com>
+//
+use crate::jsonbuilder::{JsonBuilder, JsonError};
+use crate::mysql::mysql::*;
+
+fn log_mysql(tx: &MysqlTransaction, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("mysql")?;
+    if let Some(version) = &tx.version {
```
Comment (catenacyber) [inline rust/src/mysql/logger.rs:25]: Seems strange that all fields are optional... Is it really the case ?

File: etc/schema.json [hunk 6]
```diff
@@ -2383,6 +2383,31 @@
...
+                },
+                "affected_rows": {
+                    "type": "integer"
+                },
+                "rows": {
+                    "type": "array",
+                    "optional": true,
+                    "items": {
+                        "type": "string"
```
Comment (catenacyber) [inline etc/schema.json:2406]: Could we have min_items so that we make sure we do not log an empty array ?

File: rust/src/mysql/mysql.rs [hunk 7]
```diff
@@ -0,0 +1,1131 @@
...
+    return 0;
+}
+
+export_tx_data_get!(rs_mysql_get_tx_data, MysqlTransaction);
+export_state_data_get!(rs_mysql_get_state_data, MysqlState);
+
+/// Detect
+/// Get the mysql query
+#[no_mangle]
+pub unsafe extern "C" fn SCMysqlTxGetCommand(
```
Comment (catenacyber) [inline rust/src/mysql/mysql.rs:1023]: This should be moved to a detect.rs file

File: rust/src/mysql/mysql.rs [hunk 8]
```diff
@@ -0,0 +1,1131 @@
...
+        } else {
+            return AppLayerResult::err();
+        }
+    }
+
+    let state_safe: &mut MysqlState = cast_pointer!(state, MysqlState);
+
+    if stream_slice.is_gap() {
+        state_safe.on_response_gap(stream_slice.gap_size());
+    } else if !stream_slice.is_empty() {
```
Comment (catenacyber) [inline rust/src/mysql/mysql.rs:976]: No need to recheck `!stream_slice.is_empty()`

File: rust/src/mysql/mysql.rs [hunk 9]
```diff
@@ -0,0 +1,1131 @@
...
+pub unsafe extern "C" fn rs_mysql_probing_tc(
+    _flow: *const Flow, _direction: u8, input: *const u8, input_len: u32, _rdir: *mut u8,
+) -> AppProto {
+    if input_len >= 1 && !input.is_null() {
+        let slice: &[u8] = build_slice!(input, input_len as usize);
+        match parse_handshake_request(slice) {
+            Ok(_) => return ALPROTO_MYSQL,
+            Err(nom7::Err::Incomplete(_)) => return ALPROTO_UNKNOWN,
+            Err(err) => {
+                SCLogError!("failed to probe response {:?}", err);
```
Comment (catenacyber) [inline rust/src/mysql/mysql.rs:903]: We should not this at error level, rather debug

File: rust/src/mysql/mysql.rs [hunk 10]
```diff
@@ -0,0 +1,1131 @@
...
+        self.request_gap = true;
+    }
+
+    pub fn on_response_gap(&mut self, _size: u32) {
+        self.response_gap = true;
+    }
+}
+
+/// Probe for a valid mysql message
+pub fn probe(i: &[u8]) -> IResult<&[u8], ()> {
```
Comment (catenacyber) [inline rust/src/mysql/mysql.rs:866]: Does it need to be pub ?

File: rust/src/mysql/mysql.rs [hunk 11]
```diff
@@ -0,0 +1,1131 @@
...
+                    SCLogDebug!("Response is {:?}", &response);
+                    if let Some(state) = self.response_next_state(response, flow) {
+                        self.state_progress = state;
+                    }
+                    let tx_completed = self.is_tx_completed();
+                    if let Some(tx) = self.find_or_create_tx() {
+                        if tx_completed {
+                            tx.complete = true;
+                        }
+                    }
```
Comment (catenacyber) [inline rust/src/mysql/mysql.rs:828]: Should we not create a new transaction with the response if ` self.find_or_create_tx() ` returns None ?

File: rust/src/mysql/mysql.rs [hunk 12]
```diff
@@ -0,0 +1,1131 @@
...
+    transactions: VecDeque<MysqlTransaction>,
+    request_gap: bool,
+    response_gap: bool,
+    state_progress: MysqlStateProgress,
+    tx_index_completed: usize,
+
+    client_flags: u32,
+    version: Option<String>,
+    tls: bool,
+    command_code: Option<u8>,
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/mysql/mysql.rs:149]: (rather part of a transaction ? )

File: rust/src/mysql/parser.rs [hunk 13]
```diff
@@ -0,0 +1,3447 @@
...
+#[allow(dead_code)]
+pub const MULTI_FACTOR_AUTHENTICATION: u32 = BIT_U32!(28);
+#[allow(dead_code)]
+pub const CLIENT_CAPABILITY_EXTENSION: u32 = BIT_U32!(29);
+#[allow(dead_code)]
+pub const CLIENT_SSL_VERIFY_SERVER_CERT: u32 = BIT_U32!(30);
+#[allow(dead_code)]
+pub const CLIENT_REMEMBER_OPTIONS: u32 = BIT_U32!(31);
+
+#[allow(dead_code)]
```
Comment (catenacyber) [inline rust/src/mysql/parser.rs:99]: This seems used, why do we need `allow(dead_code)]` ?

File: rust/src/mysql/parser.rs [hunk 14]
```diff
@@ -0,0 +1,3447 @@
...
+    Blob = 252,
+    VarString = 253,
+    String = 254,
+    Geometry = 255,
+}
+
+impl From<u8> for FieldType {
+    fn from(value: u8) -> Self {
+        match value {
+            0 => Self::Decimal,
```
Comment (catenacyber) [inline rust/src/mysql/parser.rs:148]: Would device StringEnum8 help here ?

File: rust/src/mysql/parser.rs [hunk 15]
```diff
@@ -0,0 +1,3447 @@
...
+    pub pkt_len: usize,
+    pub pkt_num: u8,
+
+    payload: &'a mut [u8],
+}
+
+impl<'a> Drop for MysqlPacket<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            std::mem::drop(Box::from_raw(self.payload as *mut [u8]));
```
Comment (catenacyber) [inline rust/src/mysql/parser.rs:198]: Why not use a Vec<u8> for ownership if we must ?

File: rust/src/mysql/parser.rs [hunk 16]
```diff
@@ -0,0 +1,3447 @@
...
+    pub database: Option<String>,
+    pub client_flags: u32,
+    pub client_plugin_name: Option<String>,
+    pub attributes: Option<Vec<MysqlHandshakeResponseAttribute>>,
+    pub zstd_compression_level: Option<u8>,
+}
+
+pub const SHA2_PASSWORD_REQUEST_PUBLIC_KEY: u8 = 2;
+pub const SHA2_PASSWORD_FAST_AUTH_SUCCESS: u8 = 3;
+pub const SHA2_PASSWORD_PERFORM_FULL_AUTH: u8 = 3;
```
Comment (catenacyber) [inline rust/src/mysql/parser.rs:321]: Should it be 4 ? an enum rather than a list of constants ?

File: rust/src/mysql/parser.rs [hunk 17]
```diff
@@ -0,0 +1,3447 @@
...
+        _ => Ok((i, length as u64)),
+    }
+}
+
+pub fn parse_packet_header(i: &[u8]) -> IResult<&[u8], MysqlPacket> {
+    let mut payload = Vec::new();
+    let mut payload_len: usize = 0;
+    let mut rem = i;
+    let mut pkt_num = None;
+    loop {
```
Comment (catenacyber) [inline rust/src/mysql/parser.rs:452]: Why do we loop to parse one header ?

**Comment by catenacyber** [comment]
There is an amazing amount of work in here :-)

**Comment by jufajardini** [comment]
I've adjusted the way the SV PR is passed, hopefully the step for preparing dependencies should work now.
Locally, I see that 3 of the sv tests are failing, expecting that the CI will highlight that, if it happens here, too.
For your next versions, you can follow this style. Paste the link to your SV PR in the `SV_BRANCH` option, like so:
```
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2067
```

:)

kudos on aaalll the work done so far, here!
For the CI failures, let us know if you need any guidance navigating and fixing them, ok?

File: rust/src/mysql/parser.rs [hunk 18]
```diff
@@ -0,0 +1,3447 @@
...
+        assert!(rem.is_empty());
+    }
+
+    // #[test]
+    //     let pkt: &[u8] = &[
+    //         22, 0, 0, 0, 23, 42, 2, 0, 0, 0, 1, 0, 0, 0, 0, 1, 8, 128, 16, 39, 0, 0, 0, 0, 0, 0, 5,
+    //         0, 0, 0, 25, 42, 2, 0, 0, 194, 0, 0, 0, 22, 83, 69, 76, 69, 67, 84, 32, 96, 117, 117,
+    //         105, 100, 96, 44, 96, 112, 97, 114, 101, 110, 116, 95, 105, 100, 96, 44, 96, 112, 97,
+    //         116, 104, 96, 44, 96, 110, 97, 109, 101, 96, 44, 96, 109, 101, 110, 117, 95, 108, 101,
+    //         118, 101, 108, 96, 44, 96, 104, 105, 100, 100, 101, 110, 96, 44, 96, 115, 111, 114,
```
Comment (jufajardini) [inline rust/src/mysql/parser.rs:2851]: Reminder to uncomment or remove these :P

File: rust/src/mysql/parser.rs [hunk 19]
```diff
@@ -0,0 +1,3447 @@
...
+
+    #[test]
+    fn test_parse_handshake_request() {
+        let pkt: &[u8] = &[
+            0x49, 0x00, 0x00, 0x00, 0x0a, 0x38, 0x2e, 0x34, 0x2e, 0x30, 0x00, 0x51, 0x00, 0x00,
+            0x00, 0x3e, 0x7d, 0x6a, 0x6a, 0x1a, 0x2d, 0x2b, 0x6b, 0x00, 0xff, 0xff, 0xff, 0x02,
+            0x00, 0xff, 0xdf, 0x15, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x61, 0x74, 0x54, 0x07, 0x62, 0x28, 0x5d, 0x21, 0x06, 0x44, 0x06, 0x62, 0x00, 0x63,
+            0x61, 0x63, 0x68, 0x69, 0x6e, 0x67, 0x5f, 0x73, 0x68, 0x61, 0x32, 0x5f, 0x70, 0x61,
+            0x73, 0x73, 0x77, 0x6f, 0x72, 0x64, 0x00,
```
Comment (jufajardini) [inline rust/src/mysql/parser.rs:2172]: As a good practice, when we have buffers of hex, especially long ones like that, could you add a comment indicating what is the buffer content?

File: rust/src/mysql/mysql.rs [hunk 20]
```diff
@@ -0,0 +1,1131 @@
...
+                if !tx_old.complete {
+                    tx_old.complete = true;
+                    MysqlState::set_event(tx_old, MysqlEvent::TooManyTransactions);
+                    break;
+                }
+            }
+            self.tx_index_completed = index;
+        }
+        tx
+    }
```
Comment (jufajardini) [inline rust/src/mysql/mysql.rs:261]: nit: add new line

**Review by jufajardini — CHANGES_REQUESTED** [review]
About the documentation:
- we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: https://github.com/OISF/suricata/commit/70ed9f91d84d)
- we should also document the log and keyword additions (one example: https://github.com/OISF/suricata/commit/70ed9f91d84d)

About the copyright note:
- all newly files should have it 
- If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?

As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!

File: suricata.yaml.in (hunk 2 above)
Comment (QianKaiLin) [reply suricata.yaml.in:923]: I copied from pgsql configuration, now i know what it means, i will remove it.

File: rust/src/mysql/mysql.rs (hunk 11 above)
Comment (QianKaiLin) [reply rust/src/mysql/mysql.rs:828]: Sorry, I don't understand the question.

File: etc/schema.json [hunk 21]
```diff
@@ -2388,23 +2388,27 @@
...
+                    "type": "string",
+                    "description": "Mysql server version"
                 },
                 "tls": {
                     "type": "boolean"
                 },
                 "command": {
-                    "type": "string"
+                    "type": "string",
+                    "description": "sql query statement or some utility commands like ping, quit etc."
```
Comment (victorjulien) [inline etc/schema.json:2399]: nit: `quit, etc.`

File: rust/src/mysql/detect.rs [hunk 22]
```diff
@@ -0,0 +1,168 @@
+/* Copyright (C) 2022 Open Information Security Foundation
```
Comment (victorjulien) [inline rust/src/mysql/detect.rs:1]: 2024

File: rust/src/mysql/detect.rs [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+#[no_mangle]
+pub unsafe extern "C" fn SCMysqlGetRowsData(
+    tx: *const c_void, _flow_flags: u8, local_id: u32, buf: *mut *const u8, len: *mut u32,
+) -> bool {
+    let tx = cast_pointer!(tx, MysqlTransaction);
+    if let Some(rows) = &tx.rows {
+        if !rows.is_empty() {
+            let index = local_id as usize;
+            if let Some(row) = rows.get(index) {
+                if !row.is_empty() {
```
Comment (victorjulien) [inline rust/src/mysql/detect.rs:118]: what does an empty row mean? Is it meaningful to match on it using `bsize:0;`?

**Review by victorjulien — CHANGES_REQUESTED** [review]
a few comments, not a full review yet

**Comment by QianKaiLin** [comment]
> About the documentation:
> 
>     * we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> 
>     * we should also document the log and keyword additions (one example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> 
> 
> About the copyright note:
> 
>     * all newly files should have it
> 
>     * If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?
> 
> 
> As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!

I have added the document in upgrade and log and keyword

**Comment by jufajardini** [comment]
> > About the documentation:
> > ```
> > * we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > 
> > * we should also document the log and keyword additions (one example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > About the copyright note:
> > ```
> > * all newly files should have it
> > 
> > * If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!
> 
> I have added the document in upgrade and log and keyword

Thanks for incorporating the feedback shared. This currently needs a rebase, due to conflicts. When you can, then, it'd be good to share a new, the most current version, rebased. Probably also after creating a new SV version :)

**Comment by QianKaiLin** [comment]
> > > About the documentation:
> > > ```
> > > * we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > > 
> > > * we should also document the log and keyword additions (one example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > About the copyright note:
> > > ```
> > > * all newly files should have it
> > > 
> > > * If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!
> > 
> > 
> > I have added the document in upgrade and log and keyword
> 
> Thanks for incorporating the feedback shared. This currently needs a rebase, due to conflicts. When you can, then, it'd be good to share a new, the most current version, rebased. Probably also after creating a new SV version :)

I have resolved conflicts, please check again, thx.

**Comment by jufajardini** [comment]
> > > > About the documentation:
> > > > ```
> > > > * we could indeed have a note on the Upgrading section (upgrading from 7 to 8, an example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > > > 
> > > > * we should also document the log and keyword additions (one example: [70ed9f91d84d](https://github.com/OISF/suricata/commit/70ed9f91d84d))
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > About the copyright note:
> > > > ```
> > > > * all newly files should have it
> > > > 
> > > > * If this work is all being done by a single author, could you please standardize the author name and e-mail used when you're adding those to the file author annotations?
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > As Philippe said, there is a massive amount of work done here, so it will take some time for us to do proper rounds of reviews. But thanks for incorporating the feedback, and the time dedicated to this!
> > > 
> > > 
> > > I have added the document in upgrade and log and keyword
> > 
> > 
> > Thanks for incorporating the feedback shared. This currently needs a rebase, due to conflicts. When you can, then, it'd be good to share a new, the most current version, rebased. Probably also after creating a new SV version :)
> 
> I have resolved conflicts, please check again, thx.

Please, submit a new PR, to make it easier for us to review the cleaned up and rebased code, as well as to ensure that the CI checks will run against a rebased and updated SV PR :)

**Pull request closed without merging**